### PR TITLE
Adds a heuristic upper bound in the case of unbounded symints

### PR DESCRIPTION
**Pull request by narendasan**
# Description

Repeated calls to torch.compile without explicit marking of dynamic dimensions may produce unbounded symints (no max val). We need a heuristic and throw a warning to unblock users. 

We now assume the upper bound is 2**16 for any particular dim (so that overflow would be less likely for tensor volume calculations done in int32). 

Added a bunch of test cases and integrated it into partitioning to generate intermediate shape ranges.

Supersedes: #4080 cc: @wenbingl 

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Review by wenbingl — APPROVED** [review]

File: py/torch_tensorrt/dynamo/partitioning/common.py [hunk 1]
```diff
@@ -26,15 +26,47 @@ def construct_dynamic_input(
...
+                min_max_opt.get(
+                    "opt",
+                    int(
+                        unwrapped_min_max_opt["min"] + unwrapped_min_max_opt["max"] / 2
+                    ),
+                )
             )
-            max_shape.append(min_max_opt["max"])
+            max_shape.append(unwrapped_min_max_opt["max"])
         else:
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/partitioning/common.py:64]: can we make this logic more consistent? something like line 54-56, make it 
```
unwrapped_min_max["opt"] = int(
                    unwrapped_min_max_opt["min"] + unwrapped_min_max_opt["max"] / 2
                )
```
and then replace line 62-67 with opt_shape.append(unwrapped_min_max["opt"]).

File: tests/py/dynamo/models/test_unbounded_symint.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+
+        def forward(self, x):
+            return F.relu(self.linear1(x))
+
+    model = SimpleModel().eval().cuda()
+
+    # Create input with unbounded batch dimension
+    input_tensor = torch.randn(4, 128).cuda()
+    # Mark dimension 0 as dynamic with only min (no max = unbounded)
+    # torch._dynamo.mark_dynamic(input_tensor, 0, min=1)
```
Comment (zewenli98) [inline tests/py/dynamo/models/test_unbounded_symint.py:40]: should this line be uncommented?

File: tests/py/dynamo/models/test_unbounded_symint.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+            self.relu = torch.nn.ReLU()
+
+        def forward(self, x):
+            return self.relu(self.conv(x))
+
+    model = ConvModel().eval().cuda()
+
+    # Create input with unbounded batch dimension
+    input_tensor = torch.randn(2, 3, 224, 224).cuda()
+    torch._dynamo.mark_dynamic(input_tensor, 0, min=1, max=10)
```
Comment (zewenli98) [inline tests/py/dynamo/models/test_unbounded_symint.py:107]: isn't it bounded?

File: tests/py/dynamo/models/test_unbounded_symint.py [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+        def forward(self, x):
+            # Simple operation that preserves shape
+            return x * 2 + 1
+
+    model = MultiDimModel().eval().cuda()
+
+    # Create input with multiple unbounded dimensions
+    input_tensor = torch.randn(4, 8, 16).cuda()
+    # Mark multiple dimensions as dynamic without max
+
```
Comment (zewenli98) [inline tests/py/dynamo/models/test_unbounded_symint.py:151]: should mark here?
